support-ast-restore-to-sql-text.md

```diff
@@ -0,0 +1,270 @@
+            {"b between '' and 'b'", "`b` BETWEEN '' AND 'b'"},
+            {"b between '2018-11-01' and '2018-11-02'", "`b` BETWEEN '2018-11-01' AND '2018-11-02'"},
+        }
+        // 为了不依赖父节点实现，通过 extractNodeFunc 抽取待测节点
+        extractNodeFunc := func(node Node) Node {
+            return node.(*SelectStmt).Fields.Fields[0].Expr
+        }
+        // Run Test
+        RunNodeRestoreTest(c, testCases, "select %s", extractNodeFunc)
+    }
+    ```
+    
+    **至此 `BetweenExpr` 的 `Restore` 函数实现完成，可以提交 PR 了。为了更好的理解测试逻辑，下面我们看 `RunNodeRestoreTest`：**
+    
+    ```
+    // 下面是测试逻辑，已经实现好了，不需要 contributor 实现
+    func RunNodeRestoreTest(c *C, nodeTestCases []NodeRestoreTestCase, template string, extractNodeFunc func(node Node) Node) {
+        parser := parser.New()
+        for _, testCase := range nodeTestCases {
+            // 通过 template 将测试用例拼接为完整的 SQL
+            sourceSQL := fmt.Sprintf(template, testCase.sourceSQL)
+            expectSQL := fmt.Sprintf(template, testCase.expectSQL)
+            stmt, err := parser.ParseOneStmt(sourceSQL, "", "")
+            comment := Commentf("source %#v", testCase)
+            c.Assert(err, IsNil, comment)
+            var sb strings.Builder
+            // 抽取指定节点并调用其 Restore 函数
+            err = extractNodeFunc(stmt).Restore(NewRestoreCtx(DefaultRestoreFlags, &sb))
+            c.Assert(err, IsNil, comment)
+            // 通过 template 将 restore 结果拼接为完整的 SQL
+            restoreSql := fmt.Sprintf(template, sb.String())
+            comment = Commentf("source %#v; restore %v", testCase, restoreSql)
+            // 测试 restore 结果与预期一致
+            c.Assert(restoreSql, Equals, expectSQL, comment)
+            stmt2, err := parser.ParseOneStmt(restoreSql, "", "")
+            c.Assert(err, IsNil, comment)
+            CleanNodeText(stmt)
+            CleanNodeText(stmt2)
+            // 测试解析的 stmt 与原 stmt 一致
+            c.Assert(stmt2, DeepEquals, stmt, comment)
+        }
+    }
+    ```
+    
+**不过对于 `ast.StmtNode`（例如：`ast.SelectStmt`）测试方法有些不一样，
+由于这类节点可以还原为一个完整的 SQL，因此直接在 `parser_test.go` 中测试。**
+
+下面以[实现 UseStmt 的 Restore 函数 PR](https://github.com/pingcap/parser/pull/62/files) 为例，对测试进行说明：
+
+1. `Restore` 函数实现过程略。
+
+2. 给函数实现添加单元测试，参见 `parser_test.go`：
+    
+    在这个示例中，只添加了几行测试数据就完成了测试：
+    
+    ```
+    // 添加 testCase 结构的测试数据
+    {"use `select`", true, "USE `select`"},
+    {"use `sel``ect`", true, "USE `sel``ect`"},
+    {"use select", false, "USE `select`"},
+    ```
+    
+    我们看 `testCase` 结构声明：
+    
+    ```
+    type testCase struct {
+        // 原 SQL
+        src     string
+        // 是否能被正确 parse
+        ok      bool
+        // 预期的 restore SQL
+        restore string
+    }
+    ```
+    
+    测试代码会判断原 SQL parse 出 AST 后再还原的 SQL 是否与预期的 restore SQL 相等，具体的测试逻辑在 `parser_test.go` 中 `RunTest()`、`RunRestoreTest()` 函数，逻辑与前例类似，此处不再赘述。
+
+---
+
+添加 TiDB Robot 微信，加入 TiDB Contributor Club，无门槛参与开源项目，改变世界从这里开始吧（萌萌哒）。
+
+![](media/tidb-robot.jpg "tidb_rpbot")
```

Comment:
This picture is better 👉 https://upload-images.jianshu.io/upload_images/542677-6952db2eff44ffe8.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240